state-chain/chains/src/sol/api.rs

```diff
@@ -148,13 +158,28 @@ impl<Environment: SolanaEnvironment> SolanaApi<Environment> {
 		let vault_program_data_account =
 			Environment::lookup_account(SolanaEnvAccountLookupKey::VaultProgramDataAccount)?;
 		let system_program_id = SolAddress::from_str(SYSTEM_PROGRAM_ID)
-			.expect("Preset System Program ID account must be valid.");
+			.expect("Preset Solana System Program ID account must be valid.");
 		let (nonce_account, durable_nonce) = Environment::nonce_account()?;
 		let compute_price = Environment::compute_price()?;
+		let token_mint_pubkey =
+			Environment::lookup_account(SolanaEnvAccountLookupKey::TokenMintPubkey)?;
+		let token_program_id = SolAddress::from_str(TOKEN_PROGRAM_ID)
+			.expect("Preset Solana Token program ID must be valid.");
+		let token_vault_ata = Environment::lookup_account(
+			SolanaEnvAccountLookupKey::TokenVaultAssociatedTokenAccount,
+		)?;
+
+		let decomposed_fetch_params = fetch_params
+			.into_iter()
+			.map(|param| AssetWithDerivedAddress::decompose_fetch_params(param, token_mint_pubkey))
+			.collect::<Result<Vec<_>, _>>()?;
 
 		// Build the instruction_set
 		let instruction_set = SolanaInstructionBuilder::fetch_from(
-			fetch_params,
+			decomposed_fetch_params,
+			token_mint_pubkey,
+			token_vault_ata,
+			token_program_id,
```

Comment:
I guess ```token_mint_pubkey``` and ```token_program_id``` (or atleast one) are Usdc related addresses and not general token realted addresses @albert-llimos? because if so, we should rename it to indicate its Usdc specific.

Reply:
`token_program_id` is the native Solana token program
`token_mint_pubkey` would be the `usdc_mint_pubkey`. I wouldn't rename anything to `usdc` because we should have the token code generic in case we decide to support new tokens in the future.

Reply:
The thing is that this is assumed to be usdc specific address in the code since when we call the ```fetch_from```, this is used to create a Usdc fetch tx so either we should make the code generic such that we read from environment in a generic way or we should rename it for now since the code treates it as it is the usdc address. This ties in to my other comment regarding us having generic code that is extendible to other sol assets. 

Also, in the CCM for tokens PR we also have a ```ccm_transfer_usdc_token``` that uses this key and there even the function name is usdc specific

Reply:
yes you're right. I'd prefer to have it generic, as everything expect the `usdc_mint_pubkey` would be exactly the same code for another token.